src/targets/gpu/jit/mlir.cpp

```diff
@@ -21,39 +21,177 @@
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  * THE SOFTWARE.
  */
-#include <migraphx/gpu/compiler.hpp>
+
+#include <iterator>
+#include <migraphx/builtin.hpp>
+#include <migraphx/instruction_ref.hpp>
+#include <migraphx/iterator_for.hpp>
 #include <migraphx/make_op.hpp>
+#include <migraphx/module.hpp>
+#include <migraphx/gpu/compiler.hpp>
 #include <migraphx/gpu/context.hpp>
+#include <migraphx/gpu/code_object_op.hpp>
 #include <migraphx/gpu/mlir.hpp>
+#include <migraphx/gpu/compile_pointwise.hpp>
 
 namespace migraphx {
 inline namespace MIGRAPHX_INLINE_NS {
 namespace gpu {
 
+static module create_pointwise_module(module_ref in_mod)
+{
+    module pw_mod;
+    std::unordered_map<instruction_ref, instruction_ref> map_ins;
+    for(auto param : in_mod->get_parameters())
+    {
+        map_ins[param] =
+            pw_mod.add_parameter(any_cast<builtin::param>(param->get_operator()).parameter,
+                                 shape{param->get_shape().type()});
+    }
+    pw_mod.add_instructions(in_mod,
+                            &map_ins,
+                            [](module& m,
+                               instruction_ref ins,
+                               const operation& op,
+                               const std::vector<instruction_ref>& inputs,
+                               const std::vector<module_ref>& mod_args) -> instruction_ref {
+                                if(op.name() == "multibroadcast" and
+                                   inputs.front()->name() == "@literal")
+                                    return inputs.front();
+                                else
+                                    return m.insert_instruction(ins, op, inputs, mod_args);
+                            });
+    return pw_mod;
+}
+
 struct mlir_compiler : compiler<mlir_compiler>
 {
     std::vector<std::string> names() const { return {"gpu::mlir_op"}; }
 
     operation compile_op(context&, const std::vector<shape>&, const value&) const { return {}; }
 
     compiler_replace
-    compile(const context& ctx, instruction_ref ins, const operation&, const value& solution) const
+    compile(context& ctx, instruction_ref ins, const operation&, const value& solution) const
     {
         auto* smod = ins->module_inputs().front();
         assert(smod->get_parameter_names().size() == ins->inputs().size() - 1);
-        return insert(compile_mlir(ctx, *smod, ins->inputs(), solution));
+        auto gemm_ins = std::find_if(smod->begin(), smod->end(), [&](const auto& i) {
+            return i.name() == "dot" or i.name() == "quant_dot";
+        });
+        // check if (a) module is fused (b) contains a dot instruction and (c) perfConfig can not
+        // allow fused module
+        if(gemm_ins != smod->end() and std::distance(gemm_ins, smod->end()) > 2 and
```

Comment:
smod->end() points to location one past  the Last instruction in the module. 
Last instruction in the fused module is a "Return". 
Therefore if module is fused with "Gemm+Pointwise" then it would be like 
```
Gemm instruction 

Pointwise instruction 

Return 

smod->end() 
```
Therefore i need to check to std::distance counts number of "hops" to reach from start to end. 
In the most simple case that i described above distance would be 3. 
Therefore it need to be >2

Reply:
That clears it up. Thanks for the breakdown. Approved